File: pagesnap_flutter/lib/services/book_store.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/book.dart';

class BookStore {
  static const String _prefix = 'book_';

  static Future<List<LocalBook>> getAll() async {
    final prefs = await SharedPreferences.getInstance();
    final keys = prefs.getKeys().where((k) => k.startsWith(_prefix));
    
    List<LocalBook> books = [];
    for (String key in keys) {
      final String? jsonStr = prefs.getString(key);
      if (jsonStr != null) {
        try {
          books.add(LocalBook.fromJson(jsonDecode(jsonStr)));
        } catch (e) {
          print('Error decoding book $key: $e');
        }
      }
    }
    
    books.sort((a, b) => b.lastRead.compareTo(a.lastRead));
    return books;
  }

  static Future<LocalBook?> getById(String id) async {
    final prefs = await SharedPreferences.getInstance();
    final jsonStr = prefs.getString('$_prefix$id');
    if (jsonStr == null) return null;
    try {
      return LocalBook.fromJson(jsonDecode(jsonStr));
    } catch (e) {
      return null;
    }
  }

  static Future<void> save(LocalBook book) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('$_prefix${book.id}', jsonEncode(book.toJson()));
  }

  static Future<void> updateProgress(String id, int wordIndex, int totalWords) async {
    final book = await getById(id);
    if (book != null) {
      book.wordIndex = wordIndex;
      book.progress = totalWords > 0 ? wordIndex / totalWords : 0;
      book.lastRead = DateTime.now().millisecondsSinceEpoch;
      await save(book);
    }
  }

  static Future<void> delete(String id) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('$_prefix$id');
  }
}

File: pagesnap-app/src/utils/BookStore.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

export interface LocalBook {
    id: string;
    title: string;
    progress: number;
    wordIndex: number;
    totalWords: number;
    lastRead: number;
    words: string[];
}

export const BookStore = {
    async getAll(): Promise<LocalBook[]> {
        try {
            const keys = await AsyncStorage.getAllKeys();
            const bookKeys = keys.filter(k => k.startsWith('book_'));
            if (bookKeys.length === 0) return [];

            const pairs = await AsyncStorage.multiGet(bookKeys);
            return pairs
                .map(([_, val]) => {
                    if (!val) return null;
                    try {
                        return JSON.parse(val) as LocalBook;
                    } catch {
                        return null;
                    }
                })
                .filter((b): b is LocalBook => b !== null)
                .sort((a, b) => b.lastRead - a.lastRead); // Most recently read first
        } catch {
            return [];
        }
    },

    async getById(id: string): Promise<LocalBook | null> {
        try {
            const json = await AsyncStorage.getItem(`book_${id}`);
            if (!json) return null;
            return JSON.parse(json) as LocalBook;
        } catch {
            return null;
        }
    },

    async save(book: LocalBook): Promise<void> {
        try {
            await AsyncStorage.setItem(`book_${book.id}`, JSON.stringify(book));
        } catch (e) {
            console.error("Failed to save book locally", e);
        }
    },

    async updateProgress(id: string, wordIndex: number, totalWords: number): Promise<void> {
        try {
            const book = await this.getById(id);
            if (!book) return;

            book.wordIndex = wordIndex;
            book.progress = totalWords > 0 ? wordIndex / totalWords : 0;
            book.lastRead = Date.now();

            await this.save(book);
        } catch (e) {
            console.error("Failed to update progress", e);
        }
    },

    async delete(id: string): Promise<void> {
        try {
            await AsyncStorage.removeItem(`book_${id}`);
        } catch (e) {
            console.error("Failed to delete book", e);
        }
    }
};
